### PR TITLE
Web log plugin

### PR DESCRIPTION
**Pull request by l2isbad**
What we have so far.
1. This plugin is replacement for nginx_log and gunicorn_log. Plugin is ok for apache (lighttpd?) access log too.
2. It is initial version. There is a lot of room for improvement.

**Charts**

- **responses** Detailed responses chart is optional. Enabled by default.
screenshot_20170208_204507

- **timings** Only if `$request_time` (nginx) or `%D` (apache) is in logs. By default chart will not appear. 
screenshot_20170208_204522

- **requests** `per url requests` chart is optional. Not enabled by default.
screenshot_20170208_204601

- **bandwidth** `received` dimension will appear only if `$request_length` (nginx) or `%I` (apache) is in logs.
screenshot_20170208_204622

- **unique clients** `All time chart `is optional. Enabled by default.
screenshot_20170208_204634

### CODE REVIEW

**Comment by ktsaou** [comment]
very nice!
I'll post here the alarms to be added, later today.

Have you tested this with apache logs?

**Comment by ktsaou** [comment]
With this apache.log:

```
::1 - - [08/Feb/2017:23:26:06 +0200] "GET /server-status?auto HTTP/1.1" 200 1239
::1 - - [08/Feb/2017:23:26:07 +0200] "GET /server-status?auto HTTP/1.1" 200 1239
::1 - - [08/Feb/2017:23:26:08 +0200] "GET /server-status?auto HTTP/1.1" 200 1239
::1 - - [08/Feb/2017:23:26:09 +0200] "GET /server-status?auto HTTP/1.1" 200 1239
::1 - - [08/Feb/2017:23:26:10 +0200] "GET /server-status?auto HTTP/1.1" 200 1239
::1 - - [08/Feb/2017:23:26:11 +0200] "GET /server-status?auto HTTP/1.1" 200 1239
::1 - - [08/Feb/2017:23:26:12 +0200] "GET /server-status?auto HTTP/1.1" 200 1240
::1 - - [08/Feb/2017:23:26:13 +0200] "GET /server-status?auto HTTP/1.1" 200 1240
::1 - - [08/Feb/2017:23:26:14 +0200] "GET /server-status?auto HTTP/1.1" 200 1240
::1 - - [08/Feb/2017:23:26:15 +0200] "GET /server-status?auto HTTP/1.1" 200 1240
```

I get:

```
2017-02-08 23:25:51: python.d ERROR: web_log_apache_log2 Can't parse output
```

I think it does not support IPv6.
Actually it would be nice to have an IPv4/IPv6 requests chart too.

**Comment by ktsaou** [comment]
Another interesting dimension in the requests chart, would be the `unmatched` lines, so that we could know it is not parsing the whole of it.

File: python.d/web_log.chart.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+from os.path import getsize
+from collections import defaultdict, namedtuple
+
+priority = 60000
+retries = 60
+
+ORDER = ['response_codes', 'request_time', 'requests_per_url', 'http_method', 'bandwidth', 'clients_cur', 'clients_all']
+CHARTS = {
+    'response_codes': {
+        'options': [None, 'Response Codes', 'requests/s', 'responses', 'web_log.response', 'stacked'],
```
Comment (ktsaou) [inline python.d/web_log.chart.py]: the context should be `web_log.response_codes`.

File: python.d/web_log.chart.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        'lines': [
+            ['2xx', '2xx', 'absolute'],
+            ['5xx', '5xx', 'absolute'],
+            ['3xx', '3xx', 'absolute'],
+            ['4xx', '4xx', 'absolute'],
+            ['1xx', '1xx', 'absolute'],
+            ['0xx', 'other', 'absolute']
+        ]},
+    'bandwidth': {
+        'options': [None, 'Bandwidth', 'KB/s', 'bandwidth', 'web_log.bw', 'area'],
```
Comment (ktsaou) [inline python.d/web_log.chart.py]: the context should be `web_log.bandwidth`.

File: python.d/web_log.chart.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        'lines': [
+            ['resp_length', 'received', 'absolute', 1, 1024],
+            ['bytes_sent', 'sent', 'absolute', -1, 1024]
+        ]},
+    'request_time': {
+        'options': [None, 'Processing Time', 'milliseconds', 'timings', 'web_log.request', 'area'],
```
Comment (ktsaou) [inline python.d/web_log.chart.py]: the context should be `web_log.request_time`

File: python.d/web_log.chart.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        'lines': [
+            ['resp_time_min', 'min', 'absolute', 1, 1],
+            ['resp_time_max', 'max', 'absolute', 1, 1],
+            ['resp_time_avg', 'avg', 'absolute', 1, 1]
+        ]},
+    'clients_cur': {
+        'options': [None, 'Current Poll Unique Client IPs', 'unique ips', 'unique clients', 'web_log.clients', 'line'],
```
Comment (ktsaou) [inline python.d/web_log.chart.py]: the context should be `web_log.clients_cur`

File: python.d/web_log.chart.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        'lines': [
+            ['unique_cur', 'clients', 'absolute', 1, 1]
+        ]},
+    'clients_all': {
+        'options': [None, 'All Time Unique Client IPs', 'unique ips', 'unique clients', 'web_log.clients', 'line'],
```
Comment (ktsaou) [inline python.d/web_log.chart.py]: the context should be `web_log.clients_all`

File: python.d/web_log.chart.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        LogService.__init__(self, configuration=configuration, name=name)
+        # Vars from module configuration file
+        self.log_path = self.configuration.get('path')
+        self.detailed_response_codes = self.configuration.get('detailed_response_codes', True)
+        self.all_time = self.configuration.get('all_time', True)
+        self.url_pattern = self.configuration.get('categories')  # dict
+        # REGEX: 1.IPv4 address 2.HTTP method 3. URL 4. Response code
+        # 5. Bytes sent 6. Response length 7. Response process time
+        #self.regex = re.compile(r'(\d{1,3}(?:\.\d{1,3}){3}) .*?"([A-Z]+) (.*?)" ([1-9]\d{2}) (\d+) (\d+)? ?([\d.]+)? ?"(.*?)" "(.*?)"')
+        self.regex = re.compile(r'(\d{1,3}(?:\.\d{1,3}){3}) .*?"([A-Z]+) (.*?)" ([1-9]\d{2}) (\d+) (\d+)? ?([\d.]+)?')
```
Comment (ktsaou) [inline python.d/web_log.chart.py]: IPv6 is not matched by this.

File: python.d/web_log.chart.py (hunk 3 above)
Comment (ktsaou) [inline python.d/web_log.chart.py]: actually, I think the chart should be named `response_time` and the context should be `web_log.response_time`.

File: python.d/web_log.chart.py (hunk 4 above)
Comment (ktsaou) [inline python.d/web_log.chart.py]: and this chart could be named `clients` and the context `web_log.clients` (as it is).

**Comment by l2isbad** [comment]
> Have you tested this with apache logs?

Not really. I check if regex matches and it matches. Will do tests later (but if regex matches we are fine).
Another "problem" with apache is `response time`
%T    The time taken to serve the request, in seconds.
or
%D    The time taken to serve the request, in microseconds.

As i say - will do real tests later.

> IPv6 is not matched by this

I know.
Can you post here some IPv6 logs?

>  (2xx, 3xx, 4xx, etc) chart, would be the unmatched lines

ok

**Comment by ktsaou** [comment]
as promised, the alarms `/etc/netdata/health.d/web_log.conf`.

@l2isbad I have used the names from my suggestions (the review comments on your code).

```

# make sure we can collect web log data

template: last_collected_secs
      on: web_log.response_codes
    calc: $now - $last_collected_t
   units: seconds ago
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ($update_every) : ( 5 * $update_every))
    crit: $this > (($status == $CRITICAL) ? ($update_every) : (60 * $update_every))
   delay: down 5m multiplier 1.5 max 1h
    info: number of seconds since the last successful data collection
      to: webmaster

# -----------------------------------------------------------------------------
# high level response code alarms

template: 1m_2xx
      on: web_log.response_codes
  lookup: sum -1m unaligned of 2xx
    calc: ($this == 0)?(1):($this)
   units: requests
   every: 10s
    info: the sum of successful HTTP requests over the last minute

template: 1m_redirects
      on: web_log.response_codes
  lookup: sum -1m unaligned of 3xx
    calc: $this * 100 / ( $1m_2xx + $this )
   units: %
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ( 1 ) : ( 2 ))
    crit: $this > (($status == $CRITICAL) ? ( 2 ) : ( 5 ))
   delay: down 15m multiplier 1.5 max 1h
    info: the ratio of HTTP redirects (3xx) vs the successful requests, \
          over the last minute
      to: webmaster

template: 1m_bad_requests
      on: web_log.response_codes
  lookup: sum -1m unaligned of 4xx
    calc: $this * 100 / ( $1m_2xx + $this )
   units: %
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ( 1 ) : (  5 ))
    crit: $this > (($status == $CRITICAL) ? ( 5 ) : ( 10 ))
   delay: down 15m multiplier 1.5 max 1h
    info: the ratio of HTTP bad requests (4xx) vs the successful requests, \
          over the last minute
      to: webmaster

template: 1m_internal_errors
      on: web_log.response_codes
  lookup: sum -1m unaligned of 5xx
    calc: $this * 100 / ( $1m_2xx + $this )
   units: %
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ( 1 ) : ( 2 ))
    crit: $this > (($status == $CRITICAL) ? ( 2 ) : ( 5 ))
   delay: down 15m multiplier 1.5 max 1h
    info: the ratio of HTTP internal server errors (5xx) vs the successful \
          requests, over the last minute
      to: webmaster

# -----------------------------------------------------------------------------
# web slow

template: 10m_response_time
      on: web_log.response_time
  lookup: average -10m unaligned of avg
   units: ms
   every: 30s
    info: the average time to respond to HTTP requests, over the last 10 minutes


template: web_slow
      on: web_log.response_time
  lookup: sum -1m unaligned of avg
   units: ms
   every: 10s
   green: 500
     red: 1000
    warn: $this > $green && $this > ($10m_response_time * 2)
    crit: $this > $red   && $this > ($10m_response_time * 4)
   delay: down 15m multiplier 1.5 max 1h
    info: the average time to respond to HTTP requests, over the last 1 minute
      to: webmaster

# -----------------------------------------------------------------------------
# web too many or too few requests

template: 5m_2xx_last
      on: web_log.response_codes
  lookup: average -5m at -5m unaligned of 2xx
   units: requests
   every: 30s
    info: average successful HTTP requests over the last 5 minutes

template: 5m_2xx_now
      on: web_log.response_codes
  lookup: average -5m unaligned of 2xx
   units: requests
   every: 30s
    info: average successful HTTP requests over the last 5 minutes

template: 5m_requests_ratio
      on: web_log.response_codes
    calc: ($5m_2xx_last > 0)?($5m_2xx_now * 100 / $5m_2xx_last):(100)
   units: %
   every: 30s
    warn: ($5m_2xx_last > 30)?($this > 200 OR $this < 50):(0)
    crit: ($5m_2xx_last > 30)?($this > 400 OR $this < 25):(0)
   delay: down 15m multiplier 1.5 max 1h
    info: the percentage of web requests over the last 5 minutes, \
          compared with the previous 15 minutes
      to: webmaster
```

**Comment by ktsaou** [comment]
> Can you post here some IPv6 logs?

I have above.

**Comment by l2isbad** [comment]
> I have above.
I saw. But it's only `::1`. How not loopback address looks like?` a-f` or `A-F` ?

**Comment by ktsaou** [comment]
Unfortunately I don't have access to one such log now, but `ifconfig` helps:

```
fe80::601:d1ff:fe16:5f01
```

**Comment by ktsaou** [comment]
another one:

```
fe80::250:8dff:feb7:e9c9
```

**Comment by ktsaou** [comment]
Check this too: https://en.wikipedia.org/wiki/IPv6

**Comment by ktsaou** [comment]
It says:

---

For convenience, an IPv6 address may be abbreviated to shorter notations by application of the following rules.

One or more leading zeroes from any groups of hexadecimal digits are removed; this is usually done to either all or none of the leading zeroes. For example, the group 0042 is converted to 42.
Consecutive sections of zeroes are replaced with a double colon (::). The double colon may only be used once in an address, as multiple use would render the address indeterminate. RFC 5952 recommends that a double colon must not be used to denote an omitted single section of zeroes.[38]
An example of application of these rules:

Initial address: 2001:0db8:0000:0000:0000:ff00:0042:8329
After removing all leading zeroes in each group: 2001:db8:0:0:0:ff00:42:8329
After omitting consecutive sections of zeroes: 2001:db8::ff00:42:8329
The loopback address, 0000:0000:0000:0000:0000:0000:0000:0001, may be abbreviated to ::1 by using both rules.

As an IPv6 address may have more than one representation, the IETF has issued a proposed standard for representing them in text.

**Comment by l2isbad** [comment]
Well, i'm totally aware about ipv6 notation in general (ISP network engineer here!). I just wanted to clarify how it looks in apache (letters are in lower or upper case)

**Comment by l2isbad** [comment]
@ktsaou line or stacked?

screenshot_20170209_105728

**Comment by l2isbad** [comment]
**Apache**:
`LogFormat "%h %l %u %t \"%r\" %>s %O %I %D \"%{Referer}i\" \"%{User-Agent}i\"" `
**default is** 
`LogFormat "%h %l %u %t \"%r\" %>s %O  \"%{Referer}i\" \"%{User-Agent}i\"" `

is equal to

**Nginx:**
```
log_format netdata '$remote_addr - $remote_user [$time_local] '
                    '"$request" $status $body_bytes_sent '
                    '$request_length $request_time '
                    '"$http_referer" "$http_user_agent"';
```

default is
```
log_format netdata '$remote_addr - $remote_user [$time_local] '
                    '"$request" $status $body_bytes_sent '
                    '"$http_referer" "$http_user_agent"';
```

**Comment by ktsaou** [comment]
> line or stacked?

stacked will allow us to have a clear view of the total.

**Comment by ktsaou** [comment]
> btw looks very similar to 12K

yes, I changed that with a PR yesterday (either 2 or 4 fractional digits)

**Comment by ktsaou** [comment]
> Well, i'm totally aware about ipv6 notation in general (big ISP network engineer here!). I just wanted to clarify how it looks in apache (letters are in lower or upper case)

Sorry... I didn't want to offend you.

I guess we should support both lower case and upper case letters.

**Comment by l2isbad** [comment]
I'm not offended 😄 
I did some tests/fixes today. Script works with apache - same set of graphs as on nginx.

**Comment by ktsaou** [comment]
> And apache auto-detection will always fail

which distro?

**Comment by l2isbad** [comment]
Debian.

**Comment by l2isbad** [comment]
@alibo I remember you wanted not aggregated detailed response codes charts.
Will do later :smile:

**Comment by alibo** [comment]
@l2isbad No problem, thanks :) 👍

**Comment by ktsaou** [comment]
merge?

**Comment by ktsaou** [comment]
oh no! can't merge.

We need a solution for disabling `nginx_log`.
@l2isbad have you thought of this?

**Comment by ktsaou** [comment]
I suggest the default config to be like this:

```yaml
# debian
nginx_log:
  name: 'nginx'
  path: '/var/log/nginx/access.log'

# gentoo
nginx_log2:
  name: 'nginx_localhost'
  path: '/var/log/nginx/localhost.access_log'

# debian
apache_log:
  name: 'apache'
  path: '/var/log/apache2/access.log'

# gentoo
apache_log2:
  name: 'apache'
  path: '/var/log/apache2/access_log'

# arch
apache_log3:
  name: 'apache'
  path: '/var/log/httpd/access_log'

apache_vhosts_log:
  name: 'apache_vhosts'
  path: '/var/log/apache2/other_vhosts_access.log'

gunicorn_log:
  name: 'gunicorn'
  path: '/var/log/gunicorn/access.log'

gunicorn_log2:
  name: 'gunicorn'
  path: '/var/log/gunicorn/gunicorn-access.log'
```

So, we need to also disable plugin `gunicorn`.

What do you think about adding user `netdata` to the `adm` group at the installer?

**Comment by l2isbad** [comment]
> We need a solution for disabling nginx_log and gunicorn_log

Yes. As i say earlier i suggest to add `nginx_log` and `gunicorn_log` to the [disabled](https://github.com/firehol/netdata/blob/master/plugins.d/python.d.plugin#L503) list and that is all. They will be disabled from start whatever in python.d.conf. Are you ok with that?

> What do you think about adding user netdata to the adm group at the installer?

Agree. Will do.

> default config to be like this

Agree. And i need  to add preferable log format to configuration file (for nginx and apache)

**Comment by ktsaou** [comment]
> Any type of requests? Just two dimensions? (Requests Per IP Version: ipv4, ipv6)

sorry I forgot this. Yes. stacked: `ipv4` and `ip6`.

> add nginx_log and gunicorn_log to the disabled list and that is all. They will be disabled from start whatever in python.d.conf. Are you ok with that?

sure, if it works. the user will not be able to enable them, but I think this will not be a problem.

File: conf.d/python.d.conf [hunk 7]
```diff
@@ -22,24 +22,36 @@ log_interval: 3600
...
+# gunicorn_log: yes
+# haproxy: yes
 # hddtemp: yes
 # ipfs: yes
 # isc_dhcpd: yes
+# mdstat: yes
 # memcached: yes
 # mysql: yes
 # nginx: yes
 # nginx_log: yes
```
Comment (ktsaou) [inline conf.d/python.d.conf]: I suggest to put it like this:

```
# nginx_log has been replaced by web_log
nginx_log: no
```

File: conf.d/python.d.conf [hunk 8]
```diff
@@ -22,24 +22,36 @@ log_interval: 3600
...
+# bind_rndc: yes
 # cpufreq: yes
+# cpuidle: yes
 # dovecot: yes
+# elasticsearch: yes
 example: no
 # exim: yes
+# fail2ban: yes
+# freeradius: yes
+# gunicorn_log: yes
```
Comment (ktsaou) [inline conf.d/python.d.conf]: as with `nginx_log`.

**Comment by ktsaou** [comment]
looks perfect. merge?

**Comment by l2isbad** [comment]
> merge?

yes!

**Comment by ktsaou** [comment]
merged! Thank you @l2isbad. Excellent job.

Please remember to update the wiki for it.
I'll add the alarms now.

**Pull request merged**